### DISCUSSION TITLE
Loader written in TypeScript not working when in node_modules

### DISCUSSION THREAD

**Discussion by inad9300** [webpack/webpack#13220] [General]
# Bug report

**What is the current behavior?**
Referencing a loader written in TypeScript works just as expected, except when the loader's TypeScript file is inside `node_modules`.

**If the current behavior is a bug, please provide the steps to reproduce.**
1. Write a simple loader using TypeScript.
2. Reference that loader from a webpack configuration file and see it working.
3. Move that loader to the `node_modules` directory and change the path in the webpack file to point to the new location.
4. See it fail with `SyntaxError: Cannot use import statement outside a module`, as webpack fails to interpret the file as TypeScript.

**What is the expected behavior?**
TypeScript webpack loaders should work even when placed in a directory named "node_modules", just the same as they work from any other directory.

**Other relevant information:**
webpack version:  5.35.1 / 4.6.0 (CLI)
Node.js version: 14.16.0
Operating System: macOS
Additional tools: ts-loader 9.1.1, ts-node 9.1.1, TypeScript 4.2.4

**Comment by alexander-akait** [comment]
> See it fail with SyntaxError: Cannot use import statement outside a module, as webpack fails to interpret the file as TypeScript.

No related to webpack

**Reply by alexander-akait** [reply]
Please provide example how you run webpack, what is `tsconfig.json` and other related information if you need help

**Reply by inad9300** [reply]
The thing is, I got a loader that works... Just not when moved into `node_modules`. I'm talking copy-paste here. When pointing to `./my-loader/my-loader.ts`, it all works as expected, but when pointing to `./node_modules/my-loader/my-loader.ts`, I get the aforementioned error. I can try proving more details if still necessary. Who's problem would you guess it is if not webpack's?

**Reply by alexander-akait** [reply]
Because webpack uses Node.js and Node.js is JavaScript engine, I think you are a little confused in something, so I suggested that you show an example of how you are doing, so that I can suggest and help.

**Reply by inad9300** [reply]
🤔 I will try to do that, though I fail to see how that would help. If webpack supports referring to loaders written in TypeScript files, I don't see how the directory these files are in should matter. webpack is able to use `my-loader.ts` when it is not inside `node_modules` just fine, despite Node.js being a JavaScript engine...

**Reply by alexander-akait** [reply]
Check you `exclude` for `ts-loader`

**Reply by inad9300** [reply]
Not sure exactly what you mean... But here is a repository reproducing the problem: https://github.com/inad9300/webpack-typescript-loader

**Reply by inad9300** [reply]
I have simplified the code further by removing `ts-loader` as a dependency. As I suspected, it has nothing to do here.

**Reply by alexander-akait** [reply]
Because you working on Node.js (it is JS engine) you need transpile your code to JS, you can't use `'./my-loader.ts'`, because webpack is not typescript and can't compile it

**Reply by inad9300** [reply]
The entire point of my GitHub issue and repository was to point out that `'./node_modules/my-loader.ts'` doesn't work, while `'./my-loader.ts'` _does_. webpack _is_ able to compile TypeScript, as demonstrated by the repository (and no less by the fact that the webpack configuration file itself is also written in TypeScript).

**Reply by alexander-akait** [reply]
Anyway it works locally. Why? Due `webpack-cli`:
1. We found your `webpack` configuration is in `ts`
2. We setup https://github.com/gulpjs/rechoir to allow using `require`/`import` with `ts` (and other) extensions, i.e. `require('./something.ts')` will work https://github.com/webpack/webpack-cli/blob/master/packages/webpack-cli/lib/webpack-cli.js#L1305
3. `rechoir` register `loader` for `ts` https://github.com/gulpjs/rechoir#what-is-it together with https://github.com/gulpjs/interpret/blob/master/index.js, it is `ts-node` (https://github.com/gulpjs/interpret/blob/master/index.js#L144)
4. You have `'./my-loader.ts'` so webpack try to loader it using `require('./my-loader.ts')` and `ts-node` return JS content (**not `ts`**).

**Do not confuse webpack loaders and Node.js registers**

But you shouldn't rely on this behavior, Node.js is JS engine. It will not work `node_modules`, because `ts-node` ignore `node_modules` directory by default

**Reply by alexander-akait** [reply]
Here line https://github.com/TypeStrong/ts-node/blob/main/dist-raw/node-cjs-loader-utils.js#L31 why it is not work

**Reply by inad9300** [reply]
If this is the case, allow me to turn the bug report into a feature request, and ask you to support loaders written in TypeScript more formally, such that people can rely on this feature. Similarly to how writing the webpack configuration file in TypeScript is massively helpful, writing webpack loaders in TypeScript would be incredibly useful too. It also gives a more cohesive picture to developers, that otherwise have to keep a mixed codebase.

Since the support is already mostly there in practice, this should be a relatively easy task, I reckon. Some of the things I see as necessary would be:
- Document this feature
- Provide type definitions for loaders (currently lacking in webpack, as far as I can tell)
- Ensure loaders can be loaded even when inside `node_modules`, or provide documentation tip on how to enable this (e.g. through [ts-node's `TS_NODE_IGNORE` environmental variable](https://github.com/TypeStrong/ts-node#cli-and-programmatic-options))

**Reply by alexander-akait** [reply]
> Ensure loaders can be loaded even when inside node_modules, or provide documentation tip on how to enable this (e.g. through ts-node's TS_NODE_IGNORE environmental variable)

webpack is not typescript compiler, sorry, we will not support this, you need compiler ts to js, please read my answer again

> Document this feature

it is not feature, it is side effect because you use `webpack.config.ts` and install `ts-node`

> Provide type definitions for loaders (currently lacking in webpack, as far as I can tell)

we are working on it https://github.com/webpack/webpack/pull/13164

**Reply by inad9300** [reply]
For what reason do you then support writing webpack configuration files in TypeScript? I find it only natural that this capability is added to loaders, extending the safety and leading to more consistency.

I have another question: have you discussed this with any of your colleagues before putting it down? I am sorry if this is not appropriate, but I feel like I haven't been taken very seriously throughout our discussion, ever since you quickly closed the original issue without even an opportunity to give an answer to your initial claim.

**Reply by alexander-akait** [reply]
Because we are not typescript compiler, please read my answer again, we can't compile ts to js, you can use `ts-node` for running your projects and webpack will support `require`/`import` for ts as you want

**Reply by inad9300** [reply]
I understand that I need to install `ts-node` so that webpack can interpret TypeScript-based configurations. I also know that I can rely on this because it is documented [here](https://webpack.js.org/configuration/configuration-languages/#typescript). And I know as well that I can directly import loaders written in TypeScript from a `webpack.config.ts` file, as I indicated already in my very first message. Yet you said "you shouldn't rely on this behavior". Which is what prompted me into asking you to turn the bug report into a feature request, because I won't use the feature if I am told that it is not to be relied upon.

With regards to my second question, should I understand the absence of an answer as an actual answer? It'd be nice to get feedback from other contributors, if possible...

**Reply by alexander-akait** [reply]
You can rely on `ts-node` behavior

>  But you shouldn't rely on this behavior, Node.js is JS engine. It will not work node_modules, because ts-node ignore node_modules directory by default

I mean when you run `webpack` make sure you use `ts-node` to make the behavior explicit

What do you want to hear from other contributors? This is how `require`/`import` works on Node.js
